File: python/uanon.helloworld.py
'''
Example 1 uses the first Tutorial. This puzzle does not include a Tezos and transaction and is not an ascension. You can access the puzzle at https://uanon.observer/tutorial/1. 

Title: Hello World
Description: The site says: {payload}
Public Key: df69b9d584c7594c819796d31b8c9b174a3c2f45f3a1e9f3443ce4831584c074
Operation: opYLmo1SAtwSWzQEMfmAmFRBucqAFZu42591XwcmgyVJrEQsUKN
Payload: Hello
Format: 1 word, uppercase first letter
'''

import hashlib
import json

# Configuration
t1_public_key = 'df69b9d584c7594c819796d31b8c9b174a3c2f45f3a1e9f3443ce4831584c074'
pad_size = 4
is_tezos_transaction = False
debug = True
debug_verbose = False

if is_tezos_transaction is True:
    hashing_rounds = 1001
else:
    hashing_rounds = 2

# Manage array of passwords
fields = ["World"]
fields = json.dumps(fields, separators=(',',':'))
len_bytes = (len(fields)).to_bytes(pad_size, byteorder='big')
fields = fields.encode()

# Data prefix
prefix = b'\x05\x01'
raw_secret = prefix + len_bytes + fields

# Rounds settings init
iterations = list(range(hashing_rounds))
last_hash = False
last_hash_b = None
attempts = []

# Hashing rounds
for i in iterations:
    if last_hash is False:
        attempts.append(raw_secret)
        hash_i = hashlib.blake2b(raw_secret, digest_size=32)
        if debug is True:
            print('Raw Secret:', raw_secret)
    else:
        hash_i = hashlib.blake2b(last_hash_b, digest_size=32)

    last_hash = hash_i.hexdigest()
    last_hash_b = hash_i.digest()
    attempts.append(last_hash)

    if debug is True:
        print('Iteration ' + str(i) + ':', last_hash)

# What is True?
assert last_hash == t1_public_key, 'Final hash does not match public key'

# The opposite of False
print('\nFinal hash (' + last_hash + ') and Public Key (' + t1_public_key + ') are equivalent\n')

File: go/uanon.helloworld.go
/**
 * Example 1 uses the first Tutorial. This puzzle does not include a Tezos and transaction and is not an ascension. You can access the puzzle at https://uanon.observer/tutorial/1. 
 * 
 * Title: Hello World
 * Description: The site says: {payload}
 * Public Key: df69b9d584c7594c819796d31b8c9b174a3c2f45f3a1e9f3443ce4831584c074
 * Operation: opYLmo1SAtwSWzQEMfmAmFRBucqAFZu42591XwcmgyVJrEQsUKN
 * Payload: Hello
 * Format: 1 word, uppercase first letter
 */

package main

import (
    "fmt"
    "testing"
    "encoding/hex"
    "encoding/json"
    "golang.org/x/crypto/blake2b"
    "github.com/stretchr/testify/assert"
)

// Configurable
const PUBLIC_KEY = "df69b9d584c7594c819796d31b8c9b174a3c2f45f3a1e9f3443ce4831584c074"
const PWD_ARRAY_SIZE = 1
const iterations = 2

func hasher(data []byte) []byte {
    res := blake2b.Sum256(data)
    return res[:]
}

func dataLength(val int) string {
    x := fmt.Sprintf("%x", val)
    for len(x) < 8 {
        x = fmt.Sprintf("0%s", x)
    }
    return x
}

func genProof(msg string, depth int) []byte {
    proof, _ := hex.DecodeString(fmt.Sprintf("0501%s%s", dataLength(len(msg)), hex.EncodeToString([]byte(msg))))
    for i := 0; i < depth; i++ {
        proof = hasher(proof)
        hexed := hex.EncodeToString(proof)
        s := fmt.Sprintf("Depth %d: %s", (i+1), hexed)
        fmt.Println(s)
    }
    return proof
}

func AssertEquals(t *testing.T, res string) {
    assert := assert.New(t)
    equality := assert.Equal(res, PUBLIC_KEY)
    if equality {
        s := fmt.Sprintf("Final hash %s and Public Key %s are equivalent", res, PUBLIC_KEY)
        fmt.Println(s)   
    } else {
        fmt.Println("Proof verification failed")
    }
}

func main() {
    // Init
    var passwords [PWD_ARRAY_SIZE]string
    passwords[0] = "World"
    json, _ := json.Marshal(passwords)
    // Generate
    s := fmt.Sprintf("Raw Secret: %s", string(json))
    fmt.Println(s)
    res := genProof(string(json), iterations)
    hexed := hex.EncodeToString(res)
    // Validate
    t := &testing.T{}
    AssertEquals(t, hexed)
}